hydra-node/test/Hydra/Network/HeartbeatSpec.hs

```diff
@@ -5,53 +5,108 @@ import Hydra.Prelude
 
-  it "sends a heartbeat message with own party id every 500 ms" $ do
+      captureIncoming receivedMessages msg =
+        atomically $ modifyTVar' receivedMessages (msg :)
+
+  it "sends a heartbeat message with own party id after 500 ms" $ do
     let sentHeartbeats = runSimOrThrow $ do
-          sentMessages <- newTVarIO ([] :: [HydraMessage SimpleTx])
+          sentMessages <- newTVarIO ([] :: [Heartbeat (HydraMessage Integer)])
 
-          withHeartbeat testHost (dummyNetwork sentMessages) noop $ \_ -> do
+          withHeartbeat 1 (captureOutgoing sentMessages) noop $ \_ ->
             threadDelay 1.1
 
           readTVarIO sentMessages
 
-    sentHeartbeats `shouldBe` replicate 2 (Ping testHost)
+    sentHeartbeats `shouldBe` [Ping 1]
+
+  it "sends Connected when Ping received from other party" $ do
+    let receivedHeartbeats = runSimOrThrow $ do
+          receivedMessages <- newTVarIO ([] :: [HydraMessage Integer])
+
+          withHeartbeat 1 (\incoming _ -> incoming (Ping 2)) (captureIncoming receivedMessages) $ \_ ->
+            threadDelay 1
+
+          readTVarIO receivedMessages
+
+    receivedHeartbeats `shouldBe` [Connected 2]
+
+  it "sends Connected when any message received from other party" $ do
+    let receivedHeartbeats = runSimOrThrow $ do
+          receivedMessages <- newTVarIO ([] :: [HydraMessage Integer])
+
+          withHeartbeat 1 (\incoming _ -> incoming (Message $ ReqTx 2 1)) (captureIncoming receivedMessages) $ \_ ->
+            threadDelay 1
+
+          readTVarIO receivedMessages
+
+    receivedHeartbeats `shouldBe` [ReqTx 2 1, Connected 2]
 
-  it "propagates Heartbeat received from other parties" $ do
-    let anotherHost = Host{hostName = "0.0.0.0", portNumber = 4001}
+  it "do not send Connected on subsequent messages from already Connected party" $ do
     let receivedHeartbeats = runSimOrThrow $ do
-          receivedMessages <- newTVarIO ([] :: [HydraMessage SimpleTx])
+          receivedMessages <- newTVarIO ([] :: [HydraMessage Integer])
 
-          let receive msg = atomically $ modifyTVar' receivedMessages (msg :)
-          withHeartbeat testHost (\cb action -> action (Network noop) >> cb (Ping anotherHost)) receive $ \_ -> do
+          withHeartbeat 1 (\incoming _ -> incoming (Message $ ReqTx 2 1) >> incoming (Ping 2)) (captureIncoming receivedMessages) $ \_ ->
             threadDelay 1
 
           readTVarIO receivedMessages
 
-    receivedHeartbeats `shouldBe` [Ping anotherHost]
+    receivedHeartbeats `shouldBe` [ReqTx 2 1, Connected 2]
+
+  it "sends Disconnected given no messages has been received from known party within twice heartbeat delay" $ do
+    let receivedHeartbeats = runSimOrThrow $ do
+          receivedMessages <- newTVarIO ([] :: [HydraMessage Integer])
+
+          let component incoming action =
+                race_
+                  (action (Network noop))
+                  (incoming (Ping 2) >> threadDelay 4 >> incoming (Ping 2) >> threadDelay 7)
```

Comment:
Having some interpretations to these numbers would've been great, e.g. `slightlyMore` and `moreThanDouble`, where these variables would also be parameterized by the actual `livenessDelay`